File: src/src/main/resources/mmlspark/JSONInputParser.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class JSONInputParser(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        headers (dict): headers of the request (default: Map())
        inputCol (str): The name of the input column
        method (str): method to use for request, (PUT, POST, PATCH) (default: POST)
        outputCol (str): The name of the output column
        url (str): Url of the service
    """

    @keyword_only
    def __init__(self, headers={}, inputCol=None, method="POST", outputCol=None, url=None):
        super(JSONInputParser, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.JSONInputParser")
        self.headers = Param(self, "headers", "headers: headers of the request (default: Map())")
        self._setDefault(headers={})
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.method = Param(self, "method", "method: method to use for request, (PUT, POST, PATCH) (default: POST)")
        self._setDefault(method="POST")
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        self.url = Param(self, "url", "url: Url of the service")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, headers={}, inputCol=None, method="POST", outputCol=None, url=None):
        """
        Set the (keyword only) parameters

        Args:

            headers (dict): headers of the request (default: Map())
            inputCol (str): The name of the input column
            method (str): method to use for request, (PUT, POST, PATCH) (default: POST)
            outputCol (str): The name of the output column
            url (str): Url of the service
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setHeaders(self, value):
        """

        Args:

            headers (dict): headers of the request (default: Map())

        """
        self._set(headers=value)
        return self


    def getHeaders(self):
        """

        Returns:

            dict: headers of the request (default: Map())
        """
        return self.getOrDefault(self.headers)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setMethod(self, value):
        """

        Args:

            method (str): method to use for request, (PUT, POST, PATCH) (default: POST)

        """
        self._set(method=value)
        return self


    def getMethod(self):
        """

        Returns:

            str: method to use for request, (PUT, POST, PATCH) (default: POST)
        """
        return self.getOrDefault(self.method)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)


    def setUrl(self, value):
        """

        Args:

            url (str): Url of the service

        """
        self._set(url=value)
        return self


    def getUrl(self):
        """

        Returns:

            str: Url of the service
        """
        return self.getOrDefault(self.url)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.JSONInputParser"

    @staticmethod
    def _from_java(java_stage):
        module_name=JSONInputParser.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".JSONInputParser"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/JSONInputParser.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Interop.Internal.Java.Util;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Io.Http
{
    /// <summary>
    /// <see cref="JSONInputParser"/> implements JSONInputParser
    /// </summary>
    public class JSONInputParser : JavaTransformer, IJavaMLWritable, IJavaMLReadable<JSONInputParser>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.io.http.JSONInputParser";

        /// <summary>
        /// Creates a <see cref="JSONInputParser"/> without any parameters.
        /// </summary>
        public JSONInputParser() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="JSONInputParser"/> with a UID that is used to give the
        /// <see cref="JSONInputParser"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public JSONInputParser(string uid) : base(s_className, uid)
        {
        }

        internal JSONInputParser(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets headers value for <see cref="headers"/>
        /// </summary>
        /// <param name="headers">
        /// headers of the request
        /// </param>
        /// <returns> New JSONInputParser object </returns>
        public JSONInputParser SetHeaders(Dictionary<string, string> value)
        {
            var hashMap = new HashMap(SparkEnvironment.JvmBridge);
            foreach (var item in value)
            {
                hashMap.Put(item.Key, item.Value);
            }
            return WrapAsJSONInputParser(Reference.Invoke("setHeaders", (object)hashMap));
        }
        
        /// <summary>
        /// Sets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <param name="inputCol">
        /// The name of the input column
        /// </param>
        /// <returns> New JSONInputParser object </returns>
        public JSONInputParser SetInputCol(string value) =>
            WrapAsJSONInputParser(Reference.Invoke("setInputCol", (object)value));
        
        /// <summary>
        /// Sets method value for <see cref="method"/>
        /// </summary>
        /// <param name="method">
        /// method to use for request, (PUT, POST, PATCH)
        /// </param>
        /// <returns> New JSONInputParser object </returns>
        public JSONInputParser SetMethod(string value) =>
            WrapAsJSONInputParser(Reference.Invoke("setMethod", (object)value));
        
        /// <summary>
        /// Sets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <param name="outputCol">
        /// The name of the output column
        /// </param>
        /// <returns> New JSONInputParser object </returns>
        public JSONInputParser SetOutputCol(string value) =>
            WrapAsJSONInputParser(Reference.Invoke("setOutputCol", (object)value));
        
        /// <summary>
        /// Sets url value for <see cref="url"/>
        /// </summary>
        /// <param name="url">
        /// Url of the service
        /// </param>
        /// <returns> New JSONInputParser object </returns>
        public JSONInputParser SetUrl(string value) =>
            WrapAsJSONInputParser(Reference.Invoke("setUrl", (object)value));

        
        /// <summary>
        /// Gets headers value for <see cref="headers"/>
        /// </summary>
        /// <returns>
        /// headers: headers of the request
        /// </returns>
        public Dictionary<string, string> GetHeaders()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getHeaders");
            JvmObjectReference hashMap = (JvmObjectReference)SparkEnvironment.JvmBridge.CallStaticJavaMethod(
                "org.apache.spark.api.dotnet.DotnetUtils", "convertToJavaMap", jvmObject);
            JvmObjectReference[] keySet = (JvmObjectReference[])(
                (JvmObjectReference)hashMap.Invoke("keySet")).Invoke("toArray");
            Dictionary<string, string> result = new Dictionary<string, string>();
            foreach (var k in keySet)
            {
                result.Add((string)k.Invoke("toString"), (string)hashMap.Invoke("get", k));
            }
            return result;
        }
        
        
        /// <summary>
        /// Gets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <returns>
        /// inputCol: The name of the input column
        /// </returns>
        public string GetInputCol() =>
            (string)Reference.Invoke("getInputCol");
        
        
        /// <summary>
        /// Gets method value for <see cref="method"/>
        /// </summary>
        /// <returns>
        /// method: method to use for request, (PUT, POST, PATCH)
        /// </returns>
        public string GetMethod() =>
            (string)Reference.Invoke("getMethod");
        
        
        /// <summary>
        /// Gets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <returns>
        /// outputCol: The name of the output column
        /// </returns>
        public string GetOutputCol() =>
            (string)Reference.Invoke("getOutputCol");
        
        
        /// <summary>
        /// Gets url value for <see cref="url"/>
        /// </summary>
        /// <returns>
        /// url: Url of the service
        /// </returns>
        public string GetUrl() =>
            (string)Reference.Invoke("getUrl");

        
        /// <summary>
        /// Loads the <see cref="JSONInputParser"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="JSONInputParser"/> was saved to</param>
        /// <returns>New <see cref="JSONInputParser"/> object, loaded from path.</returns>
        public static JSONInputParser Load(string path) => WrapAsJSONInputParser(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<JSONInputParser> Read() =>
            new JavaMLReader<JSONInputParser>((JvmObjectReference)Reference.Invoke("read"));

        private static JSONInputParser WrapAsJSONInputParser(object obj) =>
            new JSONInputParser((JvmObjectReference)obj);

        
    }
}

        